datafusion/common/src/types/logical.rs

```diff
@@ -0,0 +1,58 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use core::fmt;
+use std::{cmp::Ordering, hash::Hash, sync::Arc};
+
+use super::NativeType;
+
+/// A reference counted [`LogicalType`]
+pub type LogicalTypeRef = Arc<dyn LogicalType>;
+
+pub trait LogicalType: fmt::Debug {
+    fn native(&self) -> &NativeType;
+    fn name(&self) -> Option<&str>;
```

Comment:
Especially given this returns Option, but also because of how it's used in Eq, let's document implementation requirements here.

I suppose we will end up having code like

```rust
if type.name().is_none() {
  // it's builtin
  match type.native() {
   ....
  }
}
```

but then, if builtins have really special treatment, we really could model them better:

```rust
enum LogicalType {
  Builtin(NativeType)
  Extension(String, NativeType)
}
```

This would be strictly better than current code, but I don't think it would actually work OK. In the future type will need to provide stuff like comparison of values, or hash of values (unless they are simply resolved as function calls -- not sure how this would work for sorting, but maybe it would)

If we don't want the builtin types to have as much special treatment, then we should have the name mandatory and perhaps structured as in https://github.com/apache/datafusion/issues/11513#issuecomment-2401814374 , see also https://github.com/findepi/datafusion/blob/findepi/types/datafusion/common/src/types/type.rs and surrounding files